parsecs/src/main/scala/Parser.scala

```diff
@@ -1 +1,116 @@
 package com.jpmc.parsecs
+
+import com.jpmc.parsecs.Parser.{pure, string}
+import com.jpmc.parsecs.Result.{Failure, Success}
+
+object Parser {
+  def from[A](f: String => Result[A]): Parser[A] = new Parser[A] {
+    override def parse(input: String): Result[A] = f(input)
+  }
+
+  // you had a non exhaustive pattern match
+
+  def pred(cond: Char => Boolean): Parser[Char] = Parser.from {
+    input =>
```

Comment:
a pattern match can be used for this, `input.headOption match case Some(firstChar) => ..., None => ...`

Reply:
nice spot will try it out!